SDKS-3941: Migrate Protect from Legacy to new Ping SDK

ancheetah opened the pull request:
# JIRA Ticket

https://pingidentity.atlassian.net/browse/SDKS-3941

## Description
Moves Ping Protect module from legacy SDK using functional pattern. Improves JS Doc annotations and unit tests.

No changeset.

cerebrl reviewed (approved):
Looks good to me. I'll let @ryanbas21 be the final arbiter as most of this has to do with the usual scaffolding/infrastructure implementation.

ancheetah commented on packages/protect/src/lib/protect.ts:
```diff
@@ -0,0 +1,62 @@
+         */
+        await import('./signals-sdk.js' as string);
+      } catch (err) {
+        console.error('error loading ping signals', err); // TODO: should we throw an error here?
```
@cerebrl @ryanbas21 I copied this from the legacy sdk but left one comment here. Wondering if we should throw an error if the signals sdk fails to load. Currently if it fails then it will continue to try to initialize anyways... but that will also fail because no methods would have been added to `window._pingOneSignals`

ryanbas21 replied:
I think throwing an error here makes sense. 

@cerebrl I'm curious if maybe it makes more sense to load the signals-sdk in the closure? That way if the signals SDK fails to load, we don't actually return the returning object so there's no confusion on the api "working" at that level.

Obviously this would make the factory async, which is a bit different than what we've typically done. But I think it could be useful to error (if we do) at the creation stage, because without the signals-sdk the api is useless.

cerebrl replied:
Yeah, I think that's a good idea.

ryanbas21 commented on packages/protect/package.json:
```diff
@@ -0,0 +1,29 @@
+{
```
@ancheetah Can we mark `sideEffects` in this package.json. I believe we have to mark the signals sdk as a side effect like we do in legacy.

ryanbas21 commented on packages/protect/src/lib/protect.types.ts:
```diff
@@ -0,0 +1,105 @@
+ */
+export interface ProtectConfig {
+  /**
+   * @param {string} envId - the environment id from your PingOne tenant
```
small note: these technically are not `params`. I think the JSDoc equivalent is `property` for these.

ancheetah replied:
Ok I can change it to property. I was confused by this doc on the typescript website, not sure if it even pertains to this, but I didn't see `property` or `interface` as supported so I opted not to use them.
https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html

ancheetah replied:
the descriptions are formatted a little better with param.
```
@param consoleLogEnabled — true to enable SDK logs in the developer console. default is false
```
vs.
```
@property — {boolean} [consoleLogEnabled] - true to enable SDK logs in the developer console. default is false
```
could this be because typescript doesnt support `property`?

ryanbas21 replied:
These are just JSDoc comments, they shouldn't have any affect on typescript. its just a description of the interface/type.

This is mostly a semantic thing. 

If you were using the api here in your editor and your intellisense was triggered, you would see @param instead of @property as a comment, its just a little bit incorrect because it's a property of the object, not a param of a function.

ryanbas21 commented on packages/protect/src/lib/protect.ts:
```diff
@@ -0,0 +1,62 @@
+ * @returns {Protect} - a set of methods
+ */
+export function createProtect(config: ProtectConfig): Protect {
+  const options = config;
```
Is this assignment necessary? We could rename the above parameter to `options` if we prefer that?

ryanbas21 commented on packages/protect/package.json:
```diff
@@ -0,0 +1,30 @@
+    "url": "git+https://github.com/ForgeRock/ping-javascript-sdk.git",
+    "directory": "packages/protect"
+  },
+  "sideEffects": true,
```
I may have misled you. What I meant was something like this:

https://github.com/ForgeRock/forgerock-javascript-sdk/blob/develop/packages/ping-protect/package.json#L14

This is my reasoning/thought process so let me know if this tracks with you.

Tree shaking is the process of dead code removal. This is a static process, so bundlers are doing this based on the static code available during the uglification/minify process.

The signals SDK is not a module. It's just like an iiffe that is executed when the script loads. Additionally, it modifies the window with new functions that it uses. These modify the global scope (window) which are side effects.

So we need to tell bundlers that this file is effectful. It can't be removed. We rely on the effects produced by this in other code that.

So here we can use an array to say the files we are asking not be removed. 

Another common input in sideEffect arrays would be CSS files (in a glob syntax) because CSS also changes the dom.

Another one would be a polyfill that extends the Array prototype.

Now - truth be told, I'm not sure the code style of factory functions is very tree shakeable to begin with because the bundler can't execute the factory to determine if some of the code is used or not later on. So it must include it. 

But I could be wrong, or maybe things advanced and I'm not sure, so its worth including it anyways.

ryanbas21 commented on packages/protect/README.md:
```diff
@@ -0,0 +1,11 @@
+# protect
```
We should also update this to be a bit more descriptive. We may be able to copy over the existing readme

https://github.com/ForgeRock/forgerock-javascript-sdk/blob/develop/packages/ping-protect/README.md

and slightly modify it.

ryanbas21 reviewed (approved):
Great work!

cerebrl commented on packages/protect/src/lib/protect.ts:
```diff
@@ -0,0 +1,65 @@
+ * @param {ProtectConfig} options - the configuration options for the PingOne Signals SDK
+ * @returns {Promise<Protect>} - a set of methods to interact with the PingOne Signals SDK
+ */
+export async function createProtect(options: ProtectConfig): Promise<Protect> {
```
I'm torn on the `createProtect` name.  @ancheetah and @ryanbas21, if we use `create` as a prefix here, we are setting a precedent for our other modules. Some don't use a prefix at all. Here are some options:

- no prefix, so just `protect`
- `create<feature>`
- `init<feature>`
- `initialize<feature>`
- `return<feature>`

Sorry for being pedantic, but I just want to think though how this naming convention scales since we will have many modules that will have to adopt this convention.

ancheetah replied:
I like the first option (no prefix, just `protect`). Not a fan of `init` because the Protect/Signals SDK doesn't actually get initialized until you call start. On the other hand using it to imply a sort of 'initializing protect api' is confusing.

ancheetah commented on packages/protect/README.md:
```diff
@@ -0,0 +1,135 @@
+# Ping Protect
```
Sure I can add this. @ryanbas21 we should add another note to the ticket for publishing protect that we will need to remove this comment in addition to creating typedoc config.

ancheetah commented:
@ryanbas21 @cerebrl please review again. Added support for marketplace nodes, updated readme with important note, and changed the module name from `createProtect` to `protect`. Also I updated the `ProtectInitializeConfig` type and it was a little strange to me. Should it just extend from the `ProtectConfig` type?

cerebrl commented on packages/protect/src/lib/protect.test.ts:
```diff
@@ -0,0 +1,171 @@
+    disableTags: false,
+    disableHub: false,
+  };
+  let protectAPI: Protect | undefined;
```
I'd like to avoid relying on side-effects within the `beforeEach` and `it` functions assigning values outside of their own function blocks. Can we initialize `protect` in each `it` block?

ryanbas21 replied:
```ts
const protectAPI: Protect;

beforeEach(() => {
  protect = await protect(options);
});

afterEach(() => {
  protect = null;
})
```

I think this is what we would have to do, right? theres no way to not have the variable assignment to use in other child scopes.

I don't think we need all of the expect/asserts inside of the `getProtect` function, it would be little redundant to do this on every creation of the protect api.

We can just do this once in it's own `it` block.

cerebrl commented on packages/protect/src/lib/protect.test.ts:
```diff
@@ -0,0 +1,171 @@
+    return protectAPI;
+  }
+
+  beforeEach(async () => {
+    vi.doMock('./signals-sdk.js', () => {
+      return {
+        default: {
+          init: vi.fn(),
+          getData: vi.fn(),
+          pauseBehavioralData: vi.fn(),
+          resumeBehavioralData: vi.fn(),
+        },
+      };
+    });
+
+    if (typeof window === 'undefined') {
+      // eslint-disable-next-line @typescript-eslint/no-explicit-any
+      global.window = {} as any;
+    }
+
+    window._pingOneSignals = {
+      init: vi.fn().mockResolvedValue(undefined),
+      getData: vi.fn().mockResolvedValue('mocked-data'),
+      pauseBehavioralData: vi.fn(),
+      resumeBehavioralData: vi.fn(),
+    };
+
+    protectAPI = await getProtect(config);
+  });
+
+  afterEach(() => {
+    vi.clearAllMocks();
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    delete (window as any)._pingOneSignals;
+  });
+
+  it('should be defined', () => {
+    expect(protect).toBeDefined();
+  });
```
Anytime I see `beforeEach` and `afterEach`, I start to wonder if we are designing what's tested correctly, or we're designing the tests right. I know we may be able to reduce some repetition by doing this, but I'd much prefer to see `protect` initialized in each `it` as well as the mocks. That way, we don't rely on side-effects or shared values.

cerebrl commented on packages/protect/src/lib/protect.test.ts:
```diff
@@ -0,0 +1,171 @@
+  });
+
+  it('should error on failed signals sdk load', async () => {
+    protectAPI = undefined;
```
This kind of stuff makes me a bit uneasy. If each `it` initializes its own protect, then doing this `protectAPI = undefined` is no longer needed.

cerebrl commented on packages/protect/src/lib/protect.ts:
```diff
@@ -0,0 +1,261 @@
+    inputCb.setInputValue(value);
+  }
+
+  const protectAPI: Protect = {
+    start,
+    getData,
+    pauseBehavioralData,
+    resumeBehavioralData,
+    getPauseBehavioralData,
+    getNodeConfig,
+    getProtectType,
+    setNodeClientError,
+    setNodeInputValue,
+  };
```
I'd like to avoid this pattern where we have a big function that returns an object that isn't apparent until the very end of the parent function body. Rather, can we just return an object at the top, and inline all of these functions.

In other words:

```
// rather than this
export async function protect() {
  // ... lots of function definitions
  const api = {
    // lots of assignment
  };
  return api;
}
```

```
// I'd rather see this
export async function protect() {
  return {
    // inline functions
  };
}
```

cerebrl commented on packages/protect/src/lib/protect.ts:
```diff
@@ -0,0 +1,261 @@
+    setNodeInputValue,
+  };
+
+  try {
+    /*
+     * Load the Ping Signals SDK
+     * this automatically pollutes the window
+     * there are no exports of this module
+     */
+    await import('./signals-sdk.js' as string);
+    return protectAPI;
+  } catch (err) {
+    console.error('error loading ping signals', err);
+    throw new Error('Failed to load PingOne Signals SDK');
+  }
```
Can we move this to the top? If this fails, I'd rather just throw early, rather than late.

cerebrl commented on packages/protect/src/lib/protect.types.ts:
```diff
@@ -0,0 +1,164 @@
+  enableTrust?: boolean;
+}
+
+export interface ProtectInitializeConfig
+  extends Omit<ProtectConfig, 'envId' | 'waitForWindowLoad' | 'hubUrl' | 'externalIdentifiers'> {
+  _type: 'PingOneProtect';
+  _action: 'protect_initialize';
+  envId?: string;
+  customHost?: string;
+}
```
Ryan says, "It's fine"

cerebrl commented on packages/protect/README.md:
````diff
@@ -0,0 +1,137 @@
+The `start` method can be called at application startup, or when you receive the `PingOneProtectInitializeCallback` callback from the server. We recommend you call `start` as soon as you can to collect as much data as possible for higher accuracy.
+
+```js
+import { createProtect } from '@pingidentity/protect';
````
This import is now just `protect`, yeah?

cerebrl reviewed (approved):
This looks good. I think there's just one correction needed in the README.md file to reflect the name change in the exported function.

ancheetah commented:
updated readme with new naming for protect module and squashed commits

ryanbas21 commented on packages/protect/src/lib/protect.ts:
```diff
@@ -0,0 +1,239 @@
+          return undefined;
+        }
+
+        const protectMetadataCb = cbs.find((metadataCallback) => {
```
is there any chance that there are no `metadata` callbacks and this produces `undefined`?

ryanbas21 replied:
I don't remember, so maybe i should look into this. I maybe wrote it like this originally.

ryanbas21 replied:
nevermind I see the check below
